FormalConjectures/ForMathlib/Computability/TuringMachine/Notation.lean

```diff
@@ -0,0 +1,197 @@
+  unless c.isDigit do
+    throwError m!"Invalid write instruction: {c} is not a numeral."
+  let n := c.val - '0'.val
+  `($(Lean.Quote.quote n.toNat))
+
+/-- `Char.toStateSyntax c stateName` outputs the syntax of the constructor corresponding to `c`
+if `c` is a capital letter (i.e. something between `A` and `Z`.). For example, `A` would output
+the syntax `stateName.A` where `stateName` is the name of the type used to index sets.
+This is used when parsing the "state" component of turing machine string representations. -/
+private def Char.toStateSyntax (c : Char) (stateName : Name) (numStates : Nat) : TermElabM Term := do
+  unless c.isUpper do
+    throwError m!"Invalid state character: {c} should be between A and Z"
+  -- The convention is to non-state letters (i.e. any letter above the highest letter denoting a
+  -- state) to denote the extra halting state.
+  if c.val - 'A'.val ≥ numStates.toUInt32 then
+    `(none)
+  else
+    return Lean.mkIdent <| .str stateName c.toString
+
+/-- `Nat.toStateSyntax n stateName` outputs the syntax of the `n`-th constructor of the type used to
+index states. -/
+private def Nat.toStateSyntax (n : Nat) (stateName : Name) : TermElabM Term := do
+  if n > 25 then
+    throwError m!"Invalid state index {n}"
+  `($(Lean.mkIdent <| .str stateName Alphabet[n]!.toString))
+
+/-- `String.toStmtSyntax s stateName` parses a component of a Turing machine string representation
+and outputs the syntax of the corresponding (state, statement) pair
+(i.e. term of type `State × Stmt`). -/
+private def String.toStmtSyntax (s : String) (stateName : Name) (numStates : Nat) : TermElabM Term := do
+  let [c_symbol, c_dir, c_state] := s.data |
+    throwError m!"Invalid transition encoding: {s} should be 3 characters long."
+  if c_symbol = '-' ∧ c_dir = '-' ∧ c_state = '-' then
+    `(none)
+  else
+    let state ← c_state.toStateSyntax stateName numStates
+    let symbol ← c_symbol.toNumeralSyntax
+    let direction ← c_dir.toDirSyntax
+    let stmt ← `(Stmt.write $symbol $direction)
+    `(some ($state, $stmt))
+
+end Util
+
+/--
+Take as input a list of strings and return an array of `matchAltExpr` syntaxes
+mapping `(state, symbol)` pairs to actions of a Turing Machine,
+given by a term of type `Option (State n)`.
+-/
+def mkMachineMatchAltExpr (L : List String) (stateName : Name) (numSymbols numStates: Nat) :
+    TermElabM (Array (TSyntax ``matchAltExpr)) := do
+  /-
+  The encoding of a TM move by a (sub)string `"ABC"` works as follows:
+    - The move the TM is currently at (the dictionary's key) is determined
+      by the position of the substring
+    - `A` gives the new symbol (`0` or `1`), `B` the direction the head moves
+      (`L` or `R`) and `C` the new state (`0, ..., n-1` or `none`)
+  -/
+
+  let moves ← L.toArray.zipIdx.mapM fun (s, i) =>
+    Array.range numSymbols |>.mapM fun symbolIdx ↦ do
+      let s_first : Term ← (s.extract ⟨3 * symbolIdx⟩ ⟨3 * (symbolIdx + 1)⟩).toStmtSyntax stateName numStates
```

Comment:
Using `⟨` to construct a string position is pretty much always wrong, as it it ignores the difference between characters and utf-8-encoded bytes. You should use `String.next` to count instead.

Reply:
Done.